Sources/ArgumentParser/Completions/BashCompletionsGenerator.swift

```diff
@@ -9,257 +9,397 @@
 
-    // Start by declaring a local var for the top-level completions.
-    // Return immediately if the completion matching hasn't moved further.
-    result += "    opts=\"\(completionWords.joined(separator: " "))\"\n"
-    for line in additionalCompletions {
-      result += "    opts=\"$opts \(line)\"\n"
+            local -r cur="${2}"
+            local -r prev="${3}"
+
+            local -i positional_number
+            local -a unparsed_words=("${COMP_WORDS[@]:1:${COMP_CWORD}}")
+
+
+        """
+
+      declareTopLevelArray = "local -a "
+    } else {
+      declareTopLevelArray = ""
     }
 
-    result += """
-          if [[ $COMP_CWORD == "\(dollarOne)" ]]; then
-              COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
-              return
-          fi
+    let positionalArguments = positionalArguments
 
-      """
+    let flags = flags
+    let options = options
+    if !flags.isEmpty || !options.isEmpty {
+      result += """
+            \(declareTopLevelArray)flags=(\(flags.joined(separator: " ")))
+            \(declareTopLevelArray)options=(\(options.joined(separator: " ")))
+            \(offerFlagsOptionsFunctionName) \(positionalArguments.count)
+
+        """
+    }
 
     // Generate the case pattern-matching statements for option values.
     // If there aren't any, skip the case block altogether.
-    let optionHandlers = generateOptionHandlers(commands)
+    let optionHandlers =
+      ArgumentSet(type, visibility: .default, parent: nil)
+      .compactMap { arg -> String? in
+        let words = arg.bashCompletionWords
+        if words.isEmpty { return nil }
+
+        // Flags don't take a value, so we don't provide follow-on completions.
+        if arg.isNullary { return nil }
+
+        return """
+              \(arg.bashCompletionWords.joined(separator: "|")))
+          \(bashValueCompletion(arg).indentingEachLine(by: 8))\
+                  return
+                  ;;
+          """
+      }
+      .joined(separator: "\n")
     if !optionHandlers.isEmpty {
       result += """
-        case $prev in
-        \(optionHandlers.indentingEachLine(by: 4))
-        esac
-        """.indentingEachLine(by: 4) + "\n"
+
+            # Offer option value completions
+            # TODO: only if ${prev} matches -* & is not an option value
+            case "${prev}" in
+        \(optionHandlers)
+            esac
+
+        """
     }
 
-    // Build out completions for the subcommands.
-    if !subcommands.isEmpty {
-      // Subcommands have their own case statement that delegates out to
-      // the subcommand completion functions.
-      result += "    case ${COMP_WORDS[\(dollarOne)]} in\n"
-      for subcommand in subcommands {
-        result += """
-          (\(subcommand._commandName))
-              \(functionName)_\(subcommand._commandName) \(subcommandArgument)
-              return
-              ;;
+    if !positionalArguments.allSatisfy({ bashValueCompletion($0).isEmpty }) {
```

Comment:
What would you think about _always_ doing the `compactMap` on lines 254+, and then, when that's not empty, adding the result plus the wrapper text (instead of this check)?